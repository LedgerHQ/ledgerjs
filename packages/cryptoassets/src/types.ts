export type { CryptoCurrencyIds } from "./currencies";

/**
 *
 */
export type Unit = {
  // display name of a given unit (example: satoshi)
  name: string;
  // string to use when formatting the unit. like 'BTC' or 'USD'
  code: string;
  // number of digits after the '.'
  magnitude: number;
  // should it always print all digits even if they are 0 (usually: true for fiats, false for cryptos)
  showAllDigits?: boolean;
  // true if the code should prefix amount when formatting
  prefixCode?: boolean;
};

/**
 *
 */
type CurrencyCommon = {
  // display name of a currency
  name: string;
  // the ticker name in exchanges / countervalue apis (e.g. BTC).
  ticker: string;
  // all units of a currency (e.g. Bitcoin have bitcoin, mBTC, bit, satoshi)
  // by convention, [0] is the default and have "highest" magnitude
  units: Unit[];
  // a shorter version of code using the symbol of the currency. like Éƒ . not all cryptocurrencies have a symbol
  symbol?: string;
  // tells if countervalue need to be disabled (typically because colliding with other coins)
  disableCountervalue?: boolean;
  delisted?: boolean;
  // some countervalue will have a ticker alias
  countervalueTicker?: string;
};

/**
 *
 */
export type TokenCurrency = CurrencyCommon & {
  type: "TokenCurrency";
  id: string;
  ledgerSignature?: string;
  contractAddress: string;
  // the currency it belongs to. e.g. 'ethereum'
  parentCurrency: CryptoCurrency;
  // the type of token in the blockchain it belongs. e.g. 'erc20'
  tokenType: string;
  // indicates this is a compound token and it's "parent" erc20 have this id
  compoundFor?: string;
};

/**
 *
 */
export type FiatCurrency = CurrencyCommon & {
  type: "FiatCurrency";
};

/**
 *
 */
export type ExplorerView = {
  tx?: string;
  address?: string;
  token?: string;
};

/**
 *
 */
export type CryptoCurrency = CurrencyCommon & {
  type: "CryptoCurrency";
  // unique internal id of a crypto currency
  id: string;
  // define if a crypto is a fork from another coin. helps dealing with split/unsplit
  forkedFrom?: string;
  // name of the app as shown in the Manager
  managerAppName: string;
  // coin type according to slip44. THIS IS NOT GUARANTEED UNIQUE across currencies (e.g testnets,..)
  coinType: number;
  // the scheme name to use when formatting an URI (without the ':')
  scheme: string;
  // used for UI
  color: string;
  family: string;
  blockAvgTime?: number;
  // in seconds
  supportsSegwit?: boolean;
  supportsNativeSegwit?: boolean;
  // if defined this coin is a testnet for another crypto (id)};
  isTestnetFor?: string;
  // TODO later we could express union of types with mandatory bitcoinLikeInfo for "bitcoin" family...
  bitcoinLikeInfo?: {
    P2PKH: number;
    P2SH: number;
    XPUBVersion?: number;
    // FIXME optional as we miss some data to fill
    hasTimestamp?: boolean;
  };
  ethereumLikeInfo?: {
    chainId: number;
    networkId?: number;
    baseChain?: string;
    hardfork?: string;
  };
  explorerViews: ExplorerView[];
  terminated?: {
    link: string;
  };
  deviceTicker?: string;
};

/**
 *
 */
export type Currency = FiatCurrency | CryptoCurrency | TokenCurrency;
