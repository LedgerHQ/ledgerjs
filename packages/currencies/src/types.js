//@flow
import type { Fiat } from "./data/fiat";

export type Rate = {
  value: number,
  fiat: Fiat
};

export type Unit = {
  // display name of a given unit (exemple: satoshi)
  name: string,
  // string to use when formatting the unit. like 'BTC' or 'USD'
  code: string,
  // number of digits after the '.'
  magnitude: number,
  // a shorter version of code using the symbol of the currency. like Éƒ
  symbol?: string,
  // should it always print all digits even if they are 0 (usually: true for fiats, false for cryptos)
  showAllDigits?: boolean
};

export type FiatUnit = Unit & {
  ticker: string
};

export type UnitValue = {| value: number, unit: Unit |};

export type BitcoinJS = {|
  bech32?: string,
  bip32: {
    private: number,
    public: number
  },
  messagePrefix: string,
  pubKeyHash: number | Uint8Array,
  scriptHash: number | Uint8Array,
  wif: number
|};

export type Currency = {|
  // unique internal id of a crypto currency
  id: string,
  // coinType is an identifier specified by https://github.com/satoshilabs/slips/blob/master/slip-0044.md
  coinType: number,
  // display name of a currency
  name: string,
  // the ticker name in exchanges / countervalue apis (e.g. BTC)
  ticker: string,
  // the scheme name to use when formatting an URI (without the ':')
  scheme: string,
  // used for UI
  color: string,
  // all units of a currency (e.g. Bitcoin have bitcoin, mBTC, bit, satoshi)
  // IMPORTANT: by convention, [0] is the default and have "highest" magnitude
  units: Unit[],
  // TODO: document apiName
  apiName?: string,
  // TODO: document xpub
  xpub?: number,
  // TODO: document isSegwitSupported
  isSegwitSupported?: boolean,
  // TODO: document handleFeePerByte
  handleFeePerByte?: boolean,
  // TODO: document areTransactionTimestamped
  areTransactionTimestamped?: boolean,
  // TODO: document sigHashType
  sigHashType?: number,
  // TODO: document bitcoinjs
  bitcoinjs?: BitcoinJS
|};

export type CurrencyEntity = Currency;
