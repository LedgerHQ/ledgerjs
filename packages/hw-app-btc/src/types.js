// @flow

/**
 */
export type TransactionInput = {
  prevout: Buffer,
  script: Buffer,
  sequence: Buffer,
  tree?: Buffer,
};

/**
 */
export type TransactionOutput = {
  amount: Buffer,
  script: Buffer,
};

/**
 */
export type Transaction = {
  version: Buffer,
  inputs: TransactionInput[],
  outputs?: TransactionOutput[],
  locktime?: Buffer,
  witness?: Buffer,
  timestamp?: Buffer,
  nVersionGroupId?: Buffer,
  nExpiryHeight?: Buffer,
  extraData?: Buffer,
};
