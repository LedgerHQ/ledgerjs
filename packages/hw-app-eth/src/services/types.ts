export type LedgerEthTransactionResolution = {
  // device serialized data that contains ERC20 data (hex format)
  erc20Tokens: Array<string>;
  // device serialized data that contains NFT data (hex format)
  nfts: Array<string>;
  // device serialized data that contains external plugin data (hex format)
  externalPlugin: Array<{ payload: string; signature: string }>;
  // device serialized data that contains plugin data (hex format)
  plugin: Array<string>;
};

export type LoadConfig = {
  nftExplorerBaseURL?: string | null;
  // example of payload https://cdn.live.ledger.com/plugins/ethereum/1.json
  // fetch against an api (base url is an api that hosts /plugins/ethereum/${chainId}.json )
  // set to null will disable it
  pluginBaseURL?: string | null;
  // provide manually some extra plugins to add for the resolution (e.g. for dev purpose)
  // object will be merged with the returned value of the Ledger cdn payload
  extraPlugins?: any | null;
};

/**
 * Allows to configure precisely what the service need to resolve.
 * for instance you can set nft:true if you need clear signing on NFTs. If you set it and it is not a NFT transaction, it should still work but will do a useless service resolution.
 */
export type ResolutionConfig = {
  // NFT resolution service
  nft?: boolean;
  // external plugins resolution service (e.G. LIDO)
  externalPlugins?: boolean;
  // ERC20 resolution service (to clear sign erc20 transfers & other actions)
  erc20?: boolean;
};

export type LedgerEthTransactionService = {
  resolveTransaction: (
    rawTxHex: string,
    loadConfig: LoadConfig,
    resolutionConfig: ResolutionConfig
  ) => Promise<LedgerEthTransactionResolution>;
};
