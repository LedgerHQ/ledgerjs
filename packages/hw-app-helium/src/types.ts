import BigNumber from "bignumber.js";

export interface PaymentV1Params {
  amount: BigNumber;
  fee: BigNumber;
  nonce: number;
  payeeAddress: string;
}
