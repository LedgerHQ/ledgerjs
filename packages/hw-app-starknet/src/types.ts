export interface ResponseBase {
  errorMessage: string
  returnCode: number
}

export interface ResponseAddress extends ResponseBase {
  publicKey: Uint8Array
}

export interface ResponseVersion extends ResponseBase {
  testMode: boolean
  major: number
  minor: number
  patch: number
  deviceLocked: boolean
  targetId: string
}

export interface ResponseAppInfo extends ResponseBase {
  appName: string
  appVersion: string
  flagLen: number
  flagsValue: number
  flagRecovery: boolean
  flagSignedMcuCode: boolean
  flagOnboarded: boolean
  flagPINValidated: boolean
}

export interface ResponseSign extends ResponseBase {
  hash: Uint8Array
  r: Uint8Array,
  s: Uint8Array,
  v: number,
}
