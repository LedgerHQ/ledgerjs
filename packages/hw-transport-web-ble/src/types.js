// @flow

export type Device = Object;
export type Characteristic = Object;
