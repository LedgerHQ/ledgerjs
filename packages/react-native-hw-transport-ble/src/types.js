// @flow
export type { BleManager } from "react-native-ble-plx";
export type Device = *;
export type Characteristic = *;
