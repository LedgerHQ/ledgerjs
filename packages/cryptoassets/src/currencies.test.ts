import {
  listFiatCurrencies,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
} from "./fiats";
import {
  listTokens,
  findTokenById,
  findTokenByAddressInCurrency,
} from "./tokens";
import {
  listCryptoCurrencies,
  hasCryptoCurrencyId,
  getCryptoCurrencyById,
  findCryptoCurrency,
  findCryptoCurrencyById,
  findCryptoCurrencyByScheme,
  findCryptoCurrencyByTicker,
  findCryptoCurrencyByKeyword,
  registerCryptoCurrency,
} from "./currencies";
import { CryptoCurrency } from "./types";

test("can get currency by coin type", () => {
  expect(getCryptoCurrencyById("bitcoin")).toMatchObject({
    id: "bitcoin",
    name: "Bitcoin",
  });
  expect(getCryptoCurrencyById("litecoin")).toMatchObject({
    id: "litecoin",
    name: "Litecoin",
  });
  expect(hasCryptoCurrencyId("bitcoin")).toBe(true);
  expect(hasCryptoCurrencyId("")).toBe(false);
  expect(() => getCryptoCurrencyById("")).toThrow();
  expect(hasCryptoCurrencyId("_")).toBe(false);
  expect(() => getCryptoCurrencyById("_")).toThrow();
});

test("can find currency", () => {
  const bitcoinMatch = {
    id: "bitcoin",
    name: "Bitcoin",
  };
  expect(findCryptoCurrency((c) => c.name === "Bitcoin")).toMatchObject(
    bitcoinMatch
  );
  expect(findCryptoCurrencyById("bitcoin")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByKeyword("btc")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByKeyword("btc")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByKeyword("btc")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByTicker("BTC")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByScheme("bitcoin")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyById("_")).toBe(undefined);
  expect(findCryptoCurrencyByKeyword("_")).toBe(undefined);
  expect(findCryptoCurrencyByKeyword("_")).toBe(undefined);
  expect(findCryptoCurrencyByKeyword("_")).toBe(undefined);
  expect(findCryptoCurrencyByTicker("_")).toBe(undefined);
  expect(findCryptoCurrencyByScheme("_")).toBe(undefined);
});

test("there are some dev cryptocurrencies", () => {
  const all = listCryptoCurrencies(true);
  const prod = listCryptoCurrencies();
  expect(all).not.toBe(prod);
  expect(all.filter((a) => !a.isTestnetFor)).toMatchObject(prod);
  expect(all.length).toBeGreaterThan(prod.length);
});

test("there are some terminated cryptocurrencies", () => {
  const all = listCryptoCurrencies(false, true);
  const supported = listCryptoCurrencies();
  expect(all).not.toBe(supported);
  expect(all.filter((a) => !a.terminated)).toMatchObject(supported);
  expect(all.length).toBeGreaterThan(supported.length);
});

test("all cryptocurrencies match (by reference) the one you get by id", () => {
  for (const c of listCryptoCurrencies()) {
    expect(c).toBe(getCryptoCurrencyById(c.id));
  }
});

test("there is no testnet or terminated coin by default", () => {
  expect(listCryptoCurrencies(false, false)).toBe(listCryptoCurrencies());
  expect(listCryptoCurrencies(true, true).length).toBeGreaterThan(
    listCryptoCurrencies().length
  );

  for (const c of listCryptoCurrencies()) {
    expect(!c.terminated).toBe(true);
    expect(!c.isTestnetFor).toBe(true);
  }
});

test("all cryptocurrencies have at least one unit", () => {
  for (const c of listCryptoCurrencies()) {
    expect(c.units.length).toBeGreaterThan(0);
  }
});

test("fiats list is always the same", () => {
  expect(listFiatCurrencies()).toEqual(listFiatCurrencies());
});

test("fiats list elements are correct", () => {
  const tickers = {};

  for (const fiat of listFiatCurrencies()) {
    expect(fiat.ticker).toBeTruthy();
    expect(typeof fiat.ticker).toBe("string");
    expect(tickers[fiat.ticker]).toBeFalsy();
    expect(fiat.units.length).toBeGreaterThan(0);
    const unit = fiat.units[0];
    expect(unit.code).toBeTruthy();
    expect(typeof unit.code).toBe("string");
    expect(unit.name).toBeTruthy();
    expect(typeof unit.name).toBe("string");
    expect(unit.magnitude).toBeGreaterThan(-1);
    expect(typeof unit.magnitude).toBe("number");
    tickers[fiat.ticker] = unit;
  }
});

test("tokens are correct", () => {
  expect(listTokens().length).toBeGreaterThan(0);

  for (const token of listTokens()) {
    expect(token.ticker).toBeTruthy();
    expect(typeof token.id).toBe("string");
    expect(typeof token.name).toBe("string");

    if (token.ledgerSignature) {
      expect(typeof token.ledgerSignature).toBe("string");
    }

    expect(typeof token.tokenType).toBe("string");
    expect(typeof token.parentCurrency).toBe("object");
    expect(hasCryptoCurrencyId(token.parentCurrency.id)).toBe(true);
    expect(typeof token.ticker).toBe("string");
    expect(token.units.length).toBeGreaterThan(0);
    const unit = token.units[0];
    expect(unit.code).toBeTruthy();
    expect(typeof unit.code).toBe("string");
    expect(unit.name).toBeTruthy();
    expect(typeof unit.name).toBe("string");
    expect(unit.magnitude).toBeGreaterThan(-1);
    expect(typeof unit.magnitude).toBe("number");

    if (token.compoundFor) {
      const t = findTokenById(token.compoundFor);
      expect(typeof t).toBe("object");
    }
  }
});

test("findTokenByAddressInCurrency", () => {
  expect(
    findTokenByAddressInCurrency(
      "0x111111111117dC0aa78b770fA6A738034120C302",
      "bsc"
    )
  ).toMatchObject({
    id: "bsc/bep20/1inch_token",
  });
  expect(
    findTokenByAddressInCurrency(
      "0x111111111117dC0aa78b770fA6A738034120C302",
      "ethereum"
    )
  ).toMatchObject({
    id: "ethereum/erc20/1inch_token",
  });
  expect(findTokenByAddressInCurrency("0x0", "bsc")).toBe(undefined);
  expect(
    findTokenByAddressInCurrency(
      "0x111111111117dC0aa78b770fA6A738034120C302",
      "tron"
    )
  ).toBe(undefined);
});

test("fiats list is sorted by ticker", () => {
  expect(
    listFiatCurrencies()
      .map((fiat) => fiat.ticker)
      .join(",")
  ).toEqual(
    listFiatCurrencies()
      .map((fiat) => fiat.ticker)
      .sort((a, b) => (a > b ? 1 : -1))
      .join(",")
  );
});

test("can get fiat by coin type", () => {
  expect(getFiatCurrencyByTicker("USD").units[0]).toMatchObject({
    magnitude: 2,
  });
  expect(getFiatCurrencyByTicker("EUR").units[0]).toMatchObject({
    magnitude: 2,
  });
  // this is not a fiat \o/
  expect(() => getFiatCurrencyByTicker("USDT").units[0]).toThrow();
  expect(hasFiatCurrencyTicker("USD")).toBe(true);
  expect(hasFiatCurrencyTicker("USDT")).toBe(false);
});

test("all USDT are countervalue enabled", () => {
  const tokens = listTokens().filter(
    (t) => t.ticker === "USDT" && !t.parentCurrency.isTestnetFor
  );
  expect(tokens.map((t) => t.id).sort()).toMatchSnapshot();
  expect(tokens.every((t) => t.disableCountervalue === false)).toBe(true);
});

test("can register a new coin externally", () => {
  const coinId = "mycoin";
  expect(() => getCryptoCurrencyById("mycoin")).toThrow(
    `currency with id "${coinId}" not found`
  );
  const mycoin = {
    type: "CryptoCurrency",
    id: coinId,
    coinType: 8008,
    name: "MyCoin",
    managerAppName: "MyCoin",
    ticker: "MYC",
    countervalueTicker: "MYC",
    scheme: "mycoin",
    color: "#ff0000",
    family: "mycoin",
    units: [
      {
        name: "MYC",
        code: "MYC",
        magnitude: 8,
      },
      {
        name: "SmallestUnit",
        code: "SMALLESTUNIT",
        magnitude: 0,
      },
    ],
    explorerViews: [
      {
        address: "https://mycoinexplorer.com/account/$address",
        tx: "https://mycoinexplorer.com/transaction/$hash",
        token: "https://mycoinexplorer.com/token/$contractAddress/?a=$address",
      },
    ],
  };
  registerCryptoCurrency(coinId, mycoin as CryptoCurrency);
  expect(getCryptoCurrencyById(coinId)).toEqual(mycoin);
});
