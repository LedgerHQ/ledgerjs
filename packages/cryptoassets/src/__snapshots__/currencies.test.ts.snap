// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all USDT are countervalue enabled 1`] = `
Array [
  "algorand/asa/312769",
  "ethereum/erc20/usd_tether__erc20_",
  "polygon/erc20/(pos)_tether_usd",
  "tron/trc20/TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t",
]
`;
